masonry/src/text/editor.rs

```diff
@@ -214,9 +230,75 @@ where
         }
     }
 
+    // --- MARK: IME ---
+    /// Set the IME preedit composing text.
+    ///
+    /// This starts composing. Composing is reset by calling [`PlainEditorTxn::clear_compose`].
+    /// While composing, it is a logic error to call anything other than
+    /// [`PlainEditorTxn::set_compose`] or [`PlainEditorTxn::clear_compose`].
+    ///
+    /// The preedit text replaces the current selection if this call starts composing.
+    ///
+    /// The selection is updated based on `cursor`, which contains the byte offsets relative to the
+    /// start of the preedit text. If `cursor` is `None`, the selection is collapsed to a caret in
+    /// front of the preedit text.
+    pub fn set_compose(&mut self, text: &str, cursor: Option<(usize, usize)>) {
+        debug_assert!(!text.is_empty());
+        debug_assert!(cursor.map(|cursor| cursor.1 <= text.len()).unwrap_or(true));
+
+        let start = if let Some(preedit_range) = self.editor.compose.clone() {
+            self.editor
+                .buffer
+                .replace_range(preedit_range.clone(), text);
+            preedit_range.start
+        } else {
+            if self.editor.selection.is_collapsed() {
+                self.editor
+                    .buffer
+                    .insert_str(self.editor.selection.text_range().start, text);
+            } else {
+                self.editor
+                    .buffer
+                    .replace_range(self.editor.selection.text_range(), text);
+            }
+            self.editor.selection.text_range().start
+        };
+        self.editor.compose = Some(start..start + text.len());
+        self.update_layout();
+
+        if let Some(cursor) = cursor {
+            // Select the location indicated by the IME.
+            self.editor.set_selection(Selection::new(
+                self.editor.cursor_at(start + cursor.0),
+                self.editor.cursor_at(start + cursor.1),
+            ));
+        } else {
+            // IME indicates nothing is to be selected: collapse the selection to a
+            // caret just in front of the preedit.
+            self.editor
+                .set_selection(self.editor.cursor_at(start).into());
+        }
+    }
+
+    /// Stop IME composing.
+    ///
+    /// This removes the IME preedit text.
+    pub fn clear_compose(&mut self) {
+        if let Some(preedit_range) = self.editor.compose.clone() {
+            self.editor.buffer.replace_range(preedit_range.clone(), "");
+            self.editor.compose = None;
+            self.update_layout();
+
+            self.editor
+                .set_selection(self.editor.cursor_at(preedit_range.start).into());
```

Comment:
`PlainEditor::layout_dirty` gets us part of the way there, but as you mentioned, we can't set the next selection until the layout is updated (it is forced to fit within the current layout). It might work if we have some `PlainEditor::defer_set_selection` or something similar, that simply waits with setting the selection until inside the `update_layout` method.